File: 1-hypercube.py
def encrypt(n, plain):
    r = random.randrange(0, n)
    while pgcd(r,n)!=1 :
        r = random.randrange(0, n)

    c = pow(n+1, plain, n**2)
    c *= pow(r, n, n**2)
    
    return c % n**2



def pgcd(a,b):
     if b==0:
         return a
     else:
         r=a%b
         return pgcd(b,r)

def decrypt(n, p, q, c):
    #lambda
    l = lcm(p-1, q-1)
    c = pow(c, l, n**2)
    #g = n+1
    g = pow(n+1, l, n**2)

    # c and g must be casted otherwise Sage looks for an inverse mod n that doesn't exist and it crashes.
    c = Integer(c)
    g = Integer(g)
    L_bot = (c - 1) / n
    L_top = (g - 1) / n
    return (L_bot // L_top) % n

def hyp1(m,n):
        c=[]
        for i in range(len(m)) :   
            c=c+[encrypt(n,m[i])]
            
        return c
def shyp1(c,x):
        k=1
        for i in range(len(c)) :   
            k=k*(c[i]**x[i])
        return k

File: 1-hypercube.c
#include "paillier.h"
#include "1-hypercube.h"
#include <gmp.h>

void Hyp1(mpz_t* m,mpz_t n,int N,mpz_t* c, mpz_t g, mpz_t n_square){
    
    for(int i=0;i<N;i++){
       mpz_set_ui( c[i],m[i]);
       encrypt(c[i],n,c[i] ,g,n_square);
    } 

} 

void sHyp1(mpz_t* ct,int N ,mpz_t* x,mpz_t plain, mpz_t n, mpz_t p, mpz_t q, mpz_t c, mpz_t g, mpz_t n_square, mpz_t p_minus, mpz_t q_minus, mpz_t lambda){
    for(int i=0;i<N;i++){
        mpz_pown(ct[i],ct[i],x[i]);
        mpz_mul(c,c,ct[i]);
    } 
    decrypt(plain,n,p,q,c,g,n_square,p_minus,q_minus,lambda);

} 